### PR TITLE
Add API to explain why a shard is or isn't assigned

### LINKED ISSUE #14593: Add shard allocation explain API to explain why shards are (or aren't) UNASSIGNED
# Idea<a id="sec-1" name="sec-1"></a>

Relates to a comment on [#8606](https://github.com/elastic/elasticsearch/issues/8686) and supersedes [#14405](https://github.com/elastic/elasticsearch/issues/14405).

We currently have the `/_cluster/reroute` API, that, with the `explain` and
`dry_run` parameters allow a user to manually specify an allocation command and
get back an explanation for why that shard can or can not undergo the requested
allocation. This is useful, however, it requires a user to know which node a
shard should be on, and to construct an allocation command for the shard.

I would like to build an API to answer one of the most often asked questions:
"\_Why is my shard UNASSIGNED?\_"

Instead of it being shard and node specific, I envision an API that looks like:

    GET /_cluster/allocation/explain
    {
      "index": "myindex"
      "shard": 0,
      "primary": false
    }

Which is basically asking "explain the allocation for a replica of shard 0 for
the 'myindex' index".

Here's an idea of how I'd like the response to look:

    {
        "shard": {
            "index": "myindex"
            "shard": 0,
            "primary": false
        },
        "cluster_info": {
            "nodes": {
                "nodeuuid1": {
                    "lowest_usage": {
                        "path": "/var/data1",
                        "free_bytes": 1000,
                        "used_bytes": 400,
                        "total_bytes": 1400,
                        "free_disk_percentage": "71.3%"
                        "used_disk_percentage": "28.6%"
                    },
                    "highest_usage": {
                        "path": "/var/data2",
                        "free_bytes": 1200,
                        "used_bytes": 600,
                        "total_bytes": 1800,
                        "free_disk_percentage": "66.6%"
                        "used_disk_percentage": "33.3%"
                    }
                },
                "nodeuuid2": {
                    "lowest_usage": {
                        "path": "/var/data1",
                        "free_bytes": 1000,
                        "used_bytes": 400,
                        "total_bytes": 1400,
                        "free_disk_percentage": "71.3%"
                        "used_disk_percentage": "28.6%"
                    },
                    "highest_usage": {
                        "path": "/var/data2",
                        "free_bytes": 1200,
                        "used_bytes": 600,
                        "total_bytes": 1800,
                        "free_disk_percentage": "66.6%"
                        "used_disk_percentage": "33.3%"
                    }
                },
                "nodeuuid3": {
                    "lowest_usage": {
                        "path": "/var/data1",
                        "free_bytes": 1000,
                        "used_bytes": 400,
                        "total_bytes": 1400,
                        "free_disk_percentage": "71.3%"
                        "used_disk_percentage": "28.6%"
                    },
                    "highest_usage": {
                        "path": "/var/data2",
                        "free_bytes": 1200,
                        "used_bytes": 600,
                        "total_bytes": 1800,
                        "free_disk_percentage": "66.6%"
                        "used_disk_percentage": "33.3%"
                    }
                }
            },
            "shard_sizes": {
                "[myindex][0][P]": 1228718,
                "[myindex][0][R]": 1231289,
                "[myindex][1][P]": 1248718,
                "[myindex][1][R]": 1298718,
            }
        },
        "nodes": {
            "nodeuuid1": {
                "final_decision": "NO"
                "decisions": [
                    {
                        "decider" : "same_shard",
                        "decision" : "NO",
                        "explanation" : "shard cannot be allocated on same node [JZU4UIPFQtWn34FyAH6VoQ] it already exists on"
                    },
                    {
                        "decider" : "snapshot_in_progress",
                        "decision" : "NO",
                        "explanation" : "a snapshot is in progress"
                    }
                ],
                "weight": 1.9
            }
            "nodeuuid2": {
                "final_decision": "NO"
                "decisions": [
                    {
                        "decider" : "node_version",
                        "decision" : "NO",
                        "explanation" : "target node version [1.4.0] is older than source node version [1.7.3]"
                    }
                ],
                "weight": 1.3
            }
            "nodeuuid3": {
                "final_decision": "YES"
                "decisions": [],
                "weight": 0.9
            }
        }
    }

Breaking down the parts:

## `shard`<a id="sec-1-1" name="sec-1-1"></a>

The same information passed into the request, so it is contained in the request
itself as well.

## `cluster_info`<a id="sec-1-2" name="sec-1-2"></a>

This roughly relates to <https://github.com/elastic/elasticsearch/issues/14405>,
however, I realized this API is not node-specific, so putting it in nodes' stats
API doesn't make sense. Instead, it's master-only information used for
allocation. Since this is gathered and used for allocation, it makes sense to
expose it here since it influences the final decision.

## `nodes`<a id="sec-1-3" name="sec-1-3"></a>

This is a map where each key is the node uuid (should probably include the node
name as well to be helpful). It has sub keys:

### `final_decision`<a id="sec-1-3-1" name="sec-1-3-1"></a>

A simple "YES" or "NO" for whether the shard can currently be allocated on this
node.

### `decisions`<a id="sec-1-3-2" name="sec-1-3-2"></a>

A list of all the "NO" decisions preventing allocation on this node. I could see
a flag being added to include all the "YES" decisions, but I think it should
default to showing "NO" only to prevent it being too verbose.

### `weight`<a id="sec-1-3-3" name="sec-1-3-3"></a>

I'd like to change the `ShardAllocator` interface to add the ability to "weigh"
a shard for a node, with whatever criteria it usually balances with. For
example, with the `BalancedShardsAllocator`, the weight would be the calculation
based on the number of shards for the index as well as the shard count.

I could see this being useful for answering the question "Okay, if all the
decisions where 'YES', where would this shard likely end up?".

It might be trickier to implement, but it could be added on at a later time.

# Conclusion<a id="sec-2" name="sec-2"></a>

I think this would go a long way towards helping users understand from a
cluster-level (instead of a single node-level) perspective why their shard can
or cannot be allocated.

Thoughts and feedback welcome! This is potentially a non-trivial amount of work
so I wanted to see what people thought before spending time implementing it.

### PR DESCRIPTION
**Pull request by dakrone**
This adds a new `/_cluster/allocation/explain` API that explains why a
shard can or cannot be allocated to nodes in the cluster. Additionally,
it will show where the master _desires_ to put the shard, according to
the `ShardsAllocator`.

It looks like this:

``` json
GET /_cluster/allocation/explain?pretty
{
  "index": "only-foo",
  "shard": 0,
  "primary": false
}
```

Though, you can optionally send an empty body, which means "explain the
allocation for the first unassigned shard you find".

The output when a shard is unassigned looks like this:

``` json
{
  "shard" : {
    "index" : "only-foo",
    "index_uuid" : "KnW0-zELRs6PK84l0r38ZA",
    "id" : 0,
    "primary" : false
  },
  "assigned" : false,
  "unassigned_info" : {
    "reason" : "INDEX_CREATED",
    "at" : "2016-03-22T20:04:23.620Z"
  },
  "nodes" : {
    "V-Spi0AyRZ6ZvKbaI3691w" : {
      "node_name" : "Susan Storm",
      "node_attributes" : {
        "bar" : "baz"
      },
      "final_decision" : "NO",
      "weight" : 0.06666675,
      "decisions" : [ {
        "decider" : "filter",
        "decision" : "NO",
        "explanation" : "node does not match index include filters [foo:\"bar\"]"
      } ]
    },
    "Qc6VL8c5RWaw1qXZ0Rg57g" : {
      "node_name" : "Slipstream",
      "node_attributes" : {
        "bar" : "baz",
        "foo" : "bar"
      },
      "final_decision" : "NO",
      "weight" : -1.3833332,
      "decisions" : [ {
        "decider" : "same_shard",
        "decision" : "NO",
        "explanation" : "the shard cannot be allocated on the same node id [Qc6VL8c5RWaw1qXZ0Rg57g] on which it already exists"
      } ]
    },
    "PzdyMZGXQdGhqTJHF_hGgA" : {
      "node_name" : "The Symbiote",
      "node_attributes" : { },
      "final_decision" : "NO",
      "weight" : 2.3166666,
      "decisions" : [ {
        "decider" : "filter",
        "decision" : "NO",
        "explanation" : "node does not match index include filters [foo:\"bar\"]"
      } ]
    }
  }
}
```

And when the shard _is_ assigned, the output looks like:

``` json
{
  "shard" : {
    "index" : "only-foo",
    "index_uuid" : "KnW0-zELRs6PK84l0r38ZA",
    "id" : 0,
    "primary" : true
  },
  "assigned" : true,
  "assigned_node_id" : "Qc6VL8c5RWaw1qXZ0Rg57g",
  "nodes" : {
    "V-Spi0AyRZ6ZvKbaI3691w" : {
      "node_name" : "Susan Storm",
      "node_attributes" : {
        "bar" : "baz"
      },
      "final_decision" : "NO",
      "weight" : 1.4499999,
      "decisions" : [ {
        "decider" : "filter",
        "decision" : "NO",
        "explanation" : "node does not match index include filters [foo:\"bar\"]"
      } ]
    },
    "Qc6VL8c5RWaw1qXZ0Rg57g" : {
      "node_name" : "Slipstream",
      "node_attributes" : {
        "bar" : "baz",
        "foo" : "bar"
      },
      "final_decision" : "CURRENTLY_ASSIGNED",
      "weight" : 0.0,
      "decisions" : [ {
        "decider" : "same_shard",
        "decision" : "NO",
        "explanation" : "the shard cannot be allocated on the same node id [Qc6VL8c5RWaw1qXZ0Rg57g] on which it already exists"
      } ]
    },
    "PzdyMZGXQdGhqTJHF_hGgA" : {
      "node_name" : "The Symbiote",
      "node_attributes" : { },
      "final_decision" : "NO",
      "weight" : 3.6999998,
      "decisions" : [ {
        "decider" : "filter",
        "decision" : "NO",
        "explanation" : "node does not match index include filters [foo:\"bar\"]"
      } ]
    }
  }
}
```

Only "NO" decisions are returned by default, but all decisions can be
shown by specifying the `?include_yes_decisions=true` parameter in the
request.

Resolves #14593

### CODE REVIEW

**Comment by nik9000** [comment]
>        "explanation" : "the shard cannot be allocated on the same node id [Qc6VL8c5RWaw1qXZ0Rg57g] on which it already exists"

Maybe reword this somehow - "there is already a copy of this shard assigned to this node" or something like that.

You example of when the shard is assigned still contains a NO decision which confused me. Can you explain?

File: buildSrc/src/main/resources/checkstyle_suppressions.xml [hunk 1]
```diff
@@ -326,17 +326,6 @@
   <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]cluster[/\\]routing[/\\]allocation[/\\]command[/\\]CancelAllocationCommand.java" checks="LineLength" />
   <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]cluster[/\\]routing[/\\]allocation[/\\]command[/\\]MoveAllocationCommand.java" checks="LineLength" />
   <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]cluster[/\\]routing[/\\]allocation[/\\]decider[/\\]AllocationDeciders.java" checks="LineLength" />
-  <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]cluster[/\\]routing[/\\]allocation[/\\]decider[/\\]AwarenessAllocationDecider.java" checks="LineLength" />
```
Comment (nik9000) [inline buildSrc/src/main/resources/checkstyle_suppressions.xml]: Hurray!

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    private boolean includeYesDecisions = false;
+
+    /** Explain the first unassigned shard */
+    public ClusterAllocationExplainRequest() {
+        this.index = null;
+        this.shard = null;
+        this.primary = null;
+    }
+
+    public ClusterAllocationExplainRequest(String index, int shard, boolean primary) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Setting primary to false here will just explain the first unassigned non-primary shard, right? Or, er, something like that. Maybe this method deserves some javadoc just to explain that.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        this.index = index;
+        return this;
+    }
+
+    @Nullable
+    public String getIndex() {
+        return this.index;
+    }
+
+    public ClusterAllocationExplainRequest shard(Integer shard) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: I think we've started to use `setShard` style here? I'm not totally sure.

**Comment by dakrone** [comment]
> Maybe reword this somehow - "there is already a copy of this shard assigned to this node" or something like that.

I'd rather leave the wording as-is for now until we determine the right thing to do in #11490, since the criteria for "same node" is likely to be changed/configurable.

> You example of when the shard is assigned still contains a NO decision which confused me. Can you explain?

Technically, the shard cannot be assigned to the node it already exists on, because it already exists there (from the allocator point of view). I thought about suppressing "NO" decisions, but decided against it because if `cluster.routing.allocation.same_shard.host` is configured, it's possible the decision may still be "YES". I opted to change the `final_decision` to `CURRENTLY_ASSIGNED` to indicate this.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        return sb.toString();
+    }
+
+    public static ClusterAllocationExplainRequest parse(XContentParser parser) throws IOException {
+        String currentFieldName = null;
+        String index = null;
+        Integer shard = null;
+        Boolean primary = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Maybe use ObjectParser?

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java (hunk 2 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Yes that's correct, if there are multiple replicas, we pick the unassigned one to explain if any are unassigned, otherwise we pick the first assigned replica. I'll add javadoc for this though to clarify that.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        extends MasterNodeOperationRequestBuilder<ClusterAllocationExplainRequest,
+                                                          ClusterAllocationExplainResponse,
+                                                          ClusterAllocationExplainRequestBuilder> {
+
+    public ClusterAllocationExplainRequestBuilder(ElasticsearchClient client, ClusterAllocationExplainAction action) {
+        super(client, action, new ClusterAllocationExplainRequest());
+    }
+
+    /** The index name to use when finding the shard to explain */
+    public ClusterAllocationExplainRequestBuilder index(String index) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java]: This to maybe is supposed to be `setIndex` now?

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A {@code ClusterAllocationExplanation} is an explanation of why a shard may or may not be allocated to nodes. It also includes weights
+ * for where the shard is likely to be assigned. It is an immutable class
+ */
+public final class ClusterAllocationExplanation implements ToXContent {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: Maybe implement Writeable and a reading constructor? That seems to be the way we're doing things more and more.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    /**
+     * Return a map of node to balancer "weight" for allocation. Higher weights mean the balancer wants to allocated the shard to that node
+     * more
+     */
+    public Map<DiscoveryNode, Float> getNodeWeights() {
+        return this.nodeWeights;
+    }
+
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(); {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: I like this style. I think I'm going to steal it.

**Comment by nik9000** [comment]
> Technically, the shard cannot be assigned to the node it already exists on, because it already exists there (from the allocator point of view). I thought about suppressing "NO" decisions, but decided against it because if cluster.routing.allocation.same_shard.host is configured, it's possible the decision may still be "YES". I opted to change the final_decision to CURRENTLY_ASSIGNED to indicate this.


OK - I'm glad it is genuinely confusing not just confusing me. I'll think on it some more and review the rest later.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java (hunk 6 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: Sure, changed this.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java (hunk 5 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java]: I changed the settings to be `setNNN` style

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java (hunk 4 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Sure, changed this to use ObjectParser

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Actually, had to revert that as `ObjectParser` doesn't correctly handle empty objects the way I want it to.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Sad!  Maybe fix it? Or we can split that to another PR if it is scary or
complicated.

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Yeah I'd prefer to handle it in a separate PR, the code in ObjectParser looks scary at first glance!

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Fine by me! Can you make an issue explaining it so we don't forget totally? I'd do it but I don't know the problem well enough.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: > the code in ObjectParser looks scary at first glance!

That's no good. Can you please keep notes as to what made you feel that way? it should be the go to place of "no problem, I can parse json settings in a breeze and with confidence"

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    /** Return turn if the shard is assigned to a node */
+    public boolean isAssigned() {
+        return this.assignedNodeId != null;
+    }
+
+    /** Return the assigned node id or null if not assigned */
+    @Nullable
+    public String assignedNodeId() {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: Probably should also be getAssignedNodeId.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+            }
+            builder.endObject(); // end shard
+            builder.field("assigned", this.assignedNodeId != null);
+            // If assigned, show the node id of the node it's assigned to
+            if (assignedNodeId != null) {
+                builder.field("assigned_node_id", this.assignedNodeId);
+            }
+            // If we have unassigned info, show that
+            if (unassignedInfo != null) {
+                builder.value(unassignedInfo);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: It feels weird to have a call to `.value` at the same level as all the calls to `.field`. I presume this works because UnassignedInfo writes its field name. Maybe call `unassignedInfo.toXContent` then? You are more of an expert on this stuff than I am though.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    public ClusterAllocationExplanation readFrom(StreamInput in) throws IOException {
+        return new ClusterAllocationExplanation(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        this.getShard().writeTo(out);
+        out.writeBoolean(this.isPrimary());
+        out.writeOptionalString(this.assignedNodeId());
+        UnassignedInfo ui = this.getUnassignedInfo();
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: Can probably be `writeOptionalWriteable` or something like that.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java [hunk 11]
```diff
@@ -215,8 +217,15 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
             // if we are above with leftover, then we know we are not good, even with mod
             if (currentNodeCount > (requiredCountPerAttribute + leftoverPerAttribute)) {
                 return allocation.decision(Decision.NO, NAME,
-                        "too many shards on node for attribute: [%s], required per attribute: [%d], node count: [%d], leftover: [%d]",
-                        awarenessAttribute, requiredCountPerAttribute, currentNodeCount, leftoverPerAttribute);
+                        "there are too many shards on the node for attribute [%s], there are [%d] total shards for the index " +
+                                " and [%d] total attributes values, expected the node count [%d] to be lower or equal to the required " +
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java]: Probably should remove the leading space here.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java [hunk 12]
```diff
@@ -63,12 +64,16 @@ private void setClusterConcurrentRebalance(int concurrentRebalance) {
     @Override
     public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
         if (clusterConcurrentRebalance == -1) {
-            return allocation.decision(Decision.YES, NAME, "all concurrent rebalances are allowed");
+            return allocation.decision(Decision.YES, NAME, "unlimited concurrent rebalances are allowed");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ConcurrentRebalanceAllocationDecider.java]: ++

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java [hunk 13]
```diff
@@ -453,14 +476,17 @@ public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAl
...
+                    "this shard is not allocated on the most utilized disk and can remain");
         }
         if (freeBytes < freeBytesThresholdHigh.bytes()) {
             if (logger.isDebugEnabled()) {
                 logger.debug("less than the required {} free bytes threshold ({} bytes free) on node {}, shard cannot remain",
                         freeBytesThresholdHigh, freeBytes, node.nodeId());
             }
-            return allocation.decision(Decision.NO, NAME, "after allocation less than required [%s] free on node, free: [%s]",
+            return allocation.decision(Decision.NO, NAME,
+                    "after allocating this shard this node would be above the high watermark " +
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java]: These are all so nice!

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java [hunk 14]
```diff
@@ -600,7 +631,7 @@ private Decision earlyTerminate(RoutingAllocation allocation, ImmutableOpenMap<S
             if (logger.isTraceEnabled()) {
                 logger.trace("cluster info unavailable for disk threshold decider, allowing allocation.");
             }
-            return allocation.decision(Decision.YES, NAME, "cluster info unavailable");
+            return allocation.decision(Decision.YES, NAME, "the cluster info unavailable");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java]: `info *is* unavailable` I think.

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/allocation/RestClusterAllocationExplainAction.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    @Inject
+    public RestClusterAllocationExplainAction(Settings settings, RestController controller, Client client) {
+        super(settings, client);
+        controller.registerHandler(RestRequest.Method.GET, "/_cluster/allocation/explain", this);
+        controller.registerHandler(RestRequest.Method.POST, "/_cluster/allocation/explain", this);
+    }
+
+    @Override
+    public void handleRequest(final RestRequest request, final RestChannel channel, final Client client) {
+        ClusterAllocationExplainRequest req = null;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/allocation/RestClusterAllocationExplainAction.java]: Null initialization feels wrong here because you will always set it. Not initializing `req` feels more right.

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/allocation/RestClusterAllocationExplainAction.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        if (RestActions.hasBodyContent(request) == false) {
+            // Empty request signals "explain the first unassigned shard you find"
+            req = new ClusterAllocationExplainRequest();
+        } else {
+            BytesReference content = RestActions.getRestContent(request);
+            try (XContentParser parser = XContentFactory.xContent(content).createParser(content)) {
+                req = ClusterAllocationExplainRequest.parse(parser);
+            } catch (Exception e) {
+                logger.error("failed to parse allocation explain request", e);
+                channel.sendResponse(new BytesRestResponse(ExceptionsHelper.status(e)));
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/allocation/RestClusterAllocationExplainAction.java]: I _think_ that you can get away with just letting the exception bubble up and RestController will send it to the user. You won't get the error log but I'm not sure logging an error on 400 level parse errors is a good thing in the long run anyway. I try to usually run requests with `error_trace` on them so we don't eat the stack trace....

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java [hunk 17]
```diff
@@ -794,7 +794,9 @@ public void addListener(Listener listener) {
             fail("should not have been able to reroute the shard");
         } catch (IllegalArgumentException e) {
             assertThat("can't allocated because there isn't enough room: " + e.getMessage(),
-                    e.getMessage().contains("more than allowed [70.0%] used disk on node, free: [26.0%]"), equalTo(true));
+                    e.getMessage()
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java]: Can you change this to assert `e.getMessage(), contains(the text)` while you are here?  That'd make the error message so much more useful.

File: rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml [hunk 18]
```diff
@@ -0,0 +1,75 @@
+---
+"cluster shard allocation explanation test":
+  - do:
+      # there aren't any unassigned shards to explain
+      catch: request
+      cluster.allocation_explain: {}
+
```
Comment (nik9000) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml]: I'd probably add `---` here and make a test case. These are difficult to debug so I take any opportunity I can to split them into smaller and smaller chunks. Totally up to you though.

File: rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+  - do:
+      indices.create:
+        index: test
+
+  - do:
```
Comment (nik9000) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml]: Maybe add a comment about how this relies on there only being a single node in the cluster. That is the case for the tests now but if someone tries to run them against a multi node cluster and doesn't understand what the API does they are going to be scratching their heads for a while.

Comment (nik9000) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml]: Actually this one doesn't require an unassigned shard. So I take that back.

Comment (nik9000) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml]: Either way these are super sensitive to the cluster layout so maybe just a single comment about that is plenty.

**Comment by nik9000** [comment]
Left some pretty innocuous questions/notes. If you want to address them and merge on your own time then LGTM.

**Comment by ywelsch** [comment]
Good work on this @dakrone (I just quickly glanced at the code, no formal review). As follow-up I would like to see an integration with `PrimaryShardAllocator` and `ReplicaShardAllocator`. This would provide insight into the following highly important scenarios (and possibly more):
- primary shard is unassigned as only stale shard copies available (allocation ids don't match)
- replica shard unassigned due to delayed shard allocation (node_left.delayed_timeout)
- primary or replica shard unassigned as shard fetching is still going on (i.e. not all data available yet to decide where to place the shard)
- replica unassigned as primary is not yet assigned

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java (hunk 9 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: the `.value` ends up calling `.toXContent`, so I don't think there's too much difference anyway? I will change it to be .toXContent though

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java (hunk 8 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: Changed this

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java (hunk 10 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java]: Sure, changed

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java (hunk 17 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java]: Sure, changed this

File: rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml (hunk 19 above)
Comment (dakrone) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cluster.allocation_explain/10_basic.yaml]: I added a comment to these tests

**Comment by dakrone** [comment]
> Good work on this @dakrone

Thanks!

> As follow-up I would like to see an integration with PrimaryShardAllocator and ReplicaShardAllocator. This would provide insight into the following highly important scenarios (and possibly more):
>  primary shard is unassigned as only stale shard copies available (allocation ids don't match)
>  replica shard unassigned due to delayed shard allocation (node_left.delayed_timeout)
>  primary or replica shard unassigned as shard fetching is still going on (i.e. not all data available yet to decide where to place the shard)

Definitely! I will open up a followup to work on integrating these things!

> replica unassigned as primary is not yet assigned

This is handled by an AllocationDecider and thus is already captured by this work :)
